011-kafka-canary.md

```diff
@@ -0,0 +1,69 @@
+# Strimzi Canary
+
+Implement a Kafka canary tool which will act as an indicator of whether Kafka clusters are operating correctly. This would be achieved by creating a canary topic and periodically producing and consuming events on the topics. 
+
+## Current situation
+
+Currently Strimzi does not have any canary style feature. It does have health checks for Kafka cluster which can be replaced/augmented with the canary.
+
+## Motivation
+
+The canary would provide the following in indication that the Kafka cluster is operating as expecting from a customers perspective. I.e. messages can successfully be produced and consumed. Some would be exported to reflect the canary's activity. Initial metrics would included:
+
+  * records-produced-total
+  * records-consumed-total
+  * produce-error
+  * consume-error
+  * latency
+
+In future the canary can be used by the Kafka Roller to improve health checks. However this is beyond of scope of the current proposed work.
+
+## Proposal
+
+The canary would be run as separate pod alongside Strimzi operator. It would be specified (optionally) in the Kafka custom resource and deployed by the Strimzi controller. The code will live in a separate repository within the Strimzi project. This is similar setup to the Kafka Bridge. 
+
+The canary will be built using Golang.
+
+There are two main options for how the topic is configured:
+
+### Option 1: Partition over replicas
+Create a topic with 1 partition per broker node and a replication factor of zero. The canary would produce/consume from each partition. 
+
+#### Pros
+
+  * Errors will indicate that a specific broker is unavailable.
+
+#### Cons
+
+  * Brokers are expected to be unavailable during rolling updates after cluster configuration changes. 
+
+### Option 2: Replicas over partition
+Create a topic with a single parition and a replication factor equal to the number of brokers and a minimum in-sync replicas of one less than the number of brokers.
```

Comment:
To make this work with scaling up and down, we will need to be able to handle somewhere the changes to the replication factor and the min-isr. The first one is currently not supported by Strimzi. So that would need to be added tot he CO. I think this is big enough chunk of work to be included in the proposal at least as a sentence or two to make it clear.

Reply:
Changing the replication factor means doing a partitions reassignment. Why not having the logic in the canary tool when it discovers that the cluster has a different number of nodes? 
We know it's something we want in Strimzi itself but it should be more related to when a user changes the replication factor on a `KafkaTopic` resource and the TO takes care of it; in this case, the canary tool doesn't interact in this way.
I would expect the canary tool to have a background worker process to check the cluster status and see if a replication factor change (aka partitions reassignment) is needed or not.

Reply:
That might be tricky especially with scale down. Will that not block the KafkaRoller?

Reply:
This is one of the concerns I wrote here https://github.com/strimzi/proposals/pull/15#issuecomment-718069412 talking about the impact on components watching the metrics (maybe the Kafka Roller).
Anyway, I heard from @tombentley that option 2 could have other issues with Kafka Roller.

Reply:
This is not about watching the metrics. This is about having a topic with 10 replicas and min.isr=9 on 5 node broker.That is why I suggested that this might need to be handled in the CO.